src/lib.rs

```diff
@@ -4,135 +4,159 @@ use core::sync::atomic;
 use core::marker::PhantomData;
 
 use std::boxed::Box;
-use std::mem;
 use std::sync::atomic::AtomicUsize;
 
 use std::sync::atomic::Ordering;
 
-pub struct Mlsp<T: ?Sized> {
-    local_count: *mut usize,
-    atomic_count: *mut AtomicUsize,
-    
-    ptr: *mut T,
-
-    phantom: PhantomData<T>
+struct Packet<T> {
+    atomic_count: AtomicUsize,
+    value: T
 }
 
-pub struct MlspPackage<T: ?Sized> {
-    atomic_count: *mut AtomicUsize,
-    
-    ptr: *mut T,
-
-    phantom: PhantomData<T>
+pub struct Mlsp<T> {
+    phantom: PhantomData<T>,
+    local_count: *mut Cell<usize>,
+    packet_ptr: *mut Packet<T>
 }
 
-impl<T> Mlsp<T> {
-    pub fn new(value: T) -> Self {
-        let local_count_cell = Cell::new(0);
-        let atomic_count_cell = Cell::new(AtomicUsize::new(0));
+pub struct MlspPackage<T> {
+    phantom: PhantomData<T>,
+    packet_ptr: *mut Packet<T>
+}
 
-        let output = Mlsp {
-            local_count: local_count_cell.as_ptr(),
-            atomic_count: atomic_count_cell.as_ptr(),
+impl<T> Packet<T> {
+    fn new(value: T) -> Self {
+        Packet {
+            atomic_count: AtomicUsize::new(1),
+            value: value
+        }
+    }
 
-            ptr: Box::into_raw(Box::new(value)),
+    fn increment(packet: *const Packet<T>) -> usize {
+        unsafe {
+            let old = (*packet).atomic_count.fetch_add(1, Ordering::Release);
+            atomic::fence(Ordering::Acquire);
```

Comment:
I can't help with atomics, as I am not that familiar with them.

Reply:
I might ask for review on the discord sometime soon